rules-tests/TypeDeclaration/Rector/Class_/TypedPropertyFromCreateMockAssignRector/Fixture/clear_false_mock.php.inc

```diff
@@ -0,0 +1,40 @@
+<?php
+
+namespace Rector\Tests\TypeDeclaration\Rector\Class_\TypedPropertyFromCreateMockAssignRector\Fixture;
+
+use PHPUnit\Framework\TestCase;
+use Rector\Tests\TypeDeclaration\Rector\Class_\TypedPropertyFromCreateMockAssignRector\Source\SomeMockedClass;
+
+final class ClearFalseMock extends TestCase
+{
+    /**
+     * @var SomeMockedClass
+     */
+    public $someMock;
+
+    protected function setUp(): void
+    {
+        $this->someMock = $this->createMock(SomeMockedClass::class);
+    }
+}
+
+?>
+-----
+<?php
+
+namespace Rector\Tests\TypeDeclaration\Rector\Class_\TypedPropertyFromCreateMockAssignRector\Fixture;
+
+use PHPUnit\Framework\TestCase;
+use Rector\Tests\TypeDeclaration\Rector\Class_\TypedPropertyFromCreateMockAssignRector\Source\SomeMockedClass;
+
+final class ClearFalseMock extends TestCase
+{
+    public \PHPUnit\Framework\MockObject\MockObject $someMock;
```

Comment:
I am thinking if it can use `@var MockObject&SomeMockedClass` as mock class is mostly intersection typed, or use intersection type if supported (8.1).

/cc @staabm any opinion?

Reply:
That would lead to error:

```
PHPDoc tag @var for property ... with type  ... is incompatible with native type                                                  PHPUnit\Framework\MockObject\MockObject.    
```

The only solution is to make `MockObject` generic, probably via phpstan/phpstan-phpunit package, but that would require having package installed.

Reply:
Another error in the `&` case is:

```bash
PHPDoc tag @var for property ... contains unresolvable type
```

Which is understandable, as class type is known and it never implements the `MockObject` interface

Reply:
what I mean is something like this:

```php
/** @var MockObject&SomeClass */
private MockObject $prop;
```

intersection should works ok with phpstan subtype, if php 8.1 already, it can be directly typed:

```php
private MockObject&MockObject $prop;
```

see PHPUnit doc https://github.com/sebastianbergmann/phpunit/blob/98b9586dbfeed26df628789440cb69eaf5595bdb/src/Framework/TestCase.php#L1338

Reply:
above is what @ruudk questioning on https://github.com/rectorphp/rector-src/pull/6177#pullrequestreview-2194458088 which valid type as I see

Reply:
Could you test it? The first case throws me an error

Reply:
I see, intersection docblock seems somehow can't resolve type when already typed `MockObject`, but union seems working:

```php
/** @var MockObject|SomeClass */
private MockObject $prop;
```

Reply:
Yet, that's not true. As it's not A or B, but always A.

Reply:
could someone post a phpstan.org/try snippet with the unexpected error?